crates/lox-core/src/math/roots.rs

```diff
@@ -243,26 +247,26 @@ impl Default for Secant {
     }
 }
 
-impl<F> FindBracketedRoot<F> for Secant
+impl<F, E> FindBracketedRoot<F, E> for Secant
 where
-    F: Fn(f64) -> f64,
+    F: Fn(f64) -> Result<f64, E>,
+    E: std::fmt::Display,
 {
-    type Error = BracketError;
-
-    fn find_in_bracket(&self, f: F, bracket: (f64, f64)) -> Result<f64, Self::Error> {
+    fn find_in_bracket(&self, f: F, bracket: (f64, f64)) -> Result<f64, RootFinderError> {
         let (x0, x1) = bracket;
         let mut p0 = x0;
         let mut p1 = x1;
-        let mut q0 = f(p0);
-        let mut q1 = f(p1);
+        let mut q0 = f(p0).map_err(|e| RootFinderError::CallbackError(e.to_string()))?;
+        let mut q1 = f(p1).map_err(|e| RootFinderError::CallbackError(e.to_string()))?;
         if q1.abs() < q0.abs() {
             std::mem::swap(&mut p0, &mut p1);
             std::mem::swap(&mut q0, &mut q1);
         }
         for i in 0..self.max_iter {
             if q1 == q0 {
                 if p1 != p0 {
-                    return Err(BracketError::NotConverged(NotConverged(i)));
+                    // XXX: unsure about p0 parameter
+                    return Err(RootFinderError::NotConverged(i, p0));
```

Comment:
The residual is the current value of the function so either `q0` or `q1` in this case. Let me check...